src/tools/base64.cpp

```diff
@@ -1,172 +1,119 @@
- */
+std::string Base64::base64encode(uint8_t const* inputString,
+                                 const size_t len)
+{
+    if ((inputString == NULL) || (len == 0 )) {
+        spdlog::error("Invalid data");
+        return std::string();
+    }
 
-#include "../tools/base64.h"
-
-#include <cstring>
-
-/* aaaack but it's fast and const should make it shared text page. */
-static const unsigned char pr2six[256] = {
-    /* ASCII table */
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 62, 64, 64, 64, 63, 52, 53, 54, 55, 56, 57, 58, 59, 60,
-    61, 64, 64, 64, 64, 64, 64, 64, 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10,
-    11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64,
-    64, 64, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42,
-    43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-    64, 64, 64, 64, 64, 64, 64, 64, 64};
-
-size_t Base64decode_len(const char* bufcoded) {
-  size_t nbytesdecoded;
-  const unsigned char* bufin;
-  int nprbytes;
-
-  bufin = reinterpret_cast<const unsigned char*>(bufcoded);
-  while (pr2six[*(bufin++)] <= 63)
-    ;
-
-  nprbytes = (bufin - (const unsigned char*)bufcoded) - 1;
-  nbytesdecoded = ((nprbytes + 3) / 4) * 3;
-
-  return nbytesdecoded + 1;
-}
+    std::string opString;
+    opString.reserve((len + 2) / 3 * 4);
+    const char* base64chars = B64chars;
+    uint32_t pos = 0;
+    constexpr int encodeWordLen = 3;
+
+    while (pos < len) {
+        opString.push_back(base64chars[(inputString[pos] & 0xfc) >> 2]);
+
+        if (pos+1 < len) {
+            opString.push_back(base64chars[((inputString[pos] & 0x03) << 4) + ((inputString[pos + 1] & 0xf0) >> 4)]);
+
+            if (pos+2 < len) {
+                opString.push_back(base64chars[((inputString[pos + 1] & 0x0f) << 2) + ((inputString[pos + 2] & 0xc0) >> 6)]);
+                opString.push_back(base64chars[inputString[pos + 2] & 0x3f]);
+            }
+            else {
+                opString.push_back(base64chars[(inputString[pos + 1] & 0x0f) << 2]);
+                opString.push_back('=');
+            }
+        }
+        else {
+            opString.push_back(base64chars[(inputString[pos] & 0x03) << 4]);
+            opString.push_back('=');
+            opString.push_back('=');
+        }
+
+        pos += encodeWordLen;
+    }
```

Comment:
When invalid array length is given (length> actual length), out of bound memory access would happen.
Add a try catch block to handle the exception.

Reply:
If we pass a length greater than actual data, the code will crash.

Reply:
Tested couple of times and didn't observe code crash, there is no option to validate actual input string length with passing length.

Reply:
As per my knowledge, it won't crash if we pass invalid data length.
Added try/catch blocks as per recommendation.

Reply:
uint8_t const* inputString , as the inputString defined as uint8_t* it means that we cannot assume null termination .

I suggest to change the definition to char* , as you are encoding a stream of characters .

in this case you can add additional sanity test (strlen(input) != len)

Please do not throw any exceptions , as the caller does not expect to catch it , and this will result with a crash

Reply:
Updated